zkevm-circuits/src/util/cell_manager_strategy.rs

```diff
@@ -0,0 +1,244 @@
+        if let Some(v) = self.0.get_mut(&cell_type) {
+            v.push(advice);
+        } else {
+            self.0.insert(cell_type, vec![advice]);
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct CMFixedWidthStrategy {
+    advices: CMFixedWidthStrategyDistribution,
+    height_offset: usize,
+
+    next: HashMap<CellType, (usize, usize)>,
+
+    perm_substitution: bool,
+}
+
+impl CMFixedWidthStrategy {
+    pub fn new(
+        advices: CMFixedWidthStrategyDistribution,
+        height_offset: usize,
+    ) -> CMFixedWidthStrategy {
+        CMFixedWidthStrategy {
+            advices,
+            height_offset,
+            next: HashMap::default(),
+            perm_substitution: false,
+        }
+    }
+
+    pub fn with_perm_substitution(mut self) -> Self {
+        self.perm_substitution = true;
+
+        self
+    }
+
+    fn get_next(&self, cell_type: &CellType) -> (usize, usize) {
+        *self.next.get(cell_type).unwrap_or(&(0, 0))
+    }
+
+    fn set_next(&mut self, cell_type: &CellType, column_idx: usize, row: usize) {
+        self.next.insert(*cell_type, (column_idx, row));
+    }
+
+    fn cells_used(&self, cell_type: &CellType, columns: &CellManagerColumns) -> usize {
+        let (next_column_idx, next_row) = self.get_next(cell_type);
+        let current_row = if next_column_idx == 0 {
+            if next_row == 0 {
+                return 0;
+            }
+
+            next_row - 1
+        } else {
+            next_row
+        };
+
+        let filled_rows_cells = if current_row == 0 {
+            0
+        } else {
+            (current_row - 1) * columns.get_cell_type_width(*cell_type)
+        };
+
+        filled_rows_cells + next_column_idx
+    }
+}
+
+impl CellManagerStrategy for CMFixedWidthStrategy {
+    fn on_creation(&mut self, columns: &mut CellManagerColumns) {
+        for (cell_type, advices) in self.advices.0.iter() {
+            for column in advices.iter() {
+                columns.add_column(*cell_type, *column)
+            }
+        }
+    }
+
+    fn query_cell<F: Field>(
+        &mut self,
+        columns: &mut CellManagerColumns,
+        meta: &mut ConstraintSystem<F>,
+        cell_type: CellType,
+    ) -> Cell<F> {
+        let (mut column_idx, mut row) = self.get_next(&cell_type);
+        if self.perm_substitution && cell_type == CellType::StoragePhase1 {
+            let (_, row_perm) = self.get_next(&CellType::StoragePermutation);
+            if row_perm < row {
+                return self.query_cell(columns, meta, CellType::StoragePermutation);
+            }
+        }
+
+        let column = columns
+            .get_column(cell_type, column_idx)
+            .expect("column not found");
+
+        let cell = Cell::new_from_cs(meta, column.advice, column.idx, self.height_offset + row);
+
+        column_idx += 1;
+        if column_idx >= columns.get_cell_type_width(cell_type) {
+            column_idx = 0;
+            row += 1;
```

Comment:
Maybe I have missed it: there should be a parameter for max height, so the number of rows doesn't grow indefinitely.  Where is it?

Reply:
For reference, this is how the old implementation dealt with this situation: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/73b2aebca9400ac8b31c9394fb726690dbfe0b74/zkevm-circuits/src/evm_circuit/util.rs#L440

Reply:
So, the max height it was necessary before, and was used only, for creating all the cells at the beggining in the constructor, now it is not necessary.

We can add a limit with assert if you think it is necessary.

Reply:
I think we should add this parameter and corresponding assertion for safety.  Extra rotations have some cost (even if it's small), and we should control how high is the highest ExecStep.  With this maximum value we will be aware of changes that require increasing the max high step, because the corresponding PR will have to update this parameter.  And to start we can use the same max value as before.